### PR TITLE
power_domain: gpio: raw follower GPIOs

### PR DESCRIPTION
**Pull request by JordanYates**
Extra list of GPIOs that 'follow' the state of the power domain. These GPIOs are set to physical level '0' when the domain transitions to SUSPENDED, and set to physical level '1' when the domain transitions to ACTIVE. The logical levels (ACTIVE_HIGH/ACTIVE_LOW) are not modified by this action (through gpio_pin_configure).

This behaviour is intended to be used when an SPI device is on a power domain and the CS pin is explicitly driven high and low by the SoC, instead of having an external pull-up to the voltage rail. The idle state of the CS pin needs to be high, but when the power domain goes down the CS pin needs to go low to avoid back-powering the device.

This property is NOT needed when the CS pins have an external pullup to the voltage rail that is being switched.

The BMI270 sensor on the Nordic Thingy53 is an example of this hardware setup.

Alternate solutions were considered:

1. Device drivers handles the problem.

Would require modification in every single SPI device driver to properly handle this, even if abstracted with a `spi_handle_power_down/up()` helper function.
 
2. Handling in SPI drivers

SPI drivers have no link to the devices, only devices link back to the SPI drivers. Even if such a link is added, the SPI drivers have no reason to be notified of power domain transitions of arbitrary devices, and don't store CS pin configuration in any case.

3. Automatic handling in the power domain driver

While the power domains end up with a list of devices on them, this is only populated at link-time, not compile-time. As a result, the power domain would need some way to query the CS pin configuration at runtime, for arbitrary devices which may or may not be SPI devices. This sounds like a thorny problem.

4. Design better hardware without this problem

Sadly the Thingy53 already exists in the world and is a Zephyr supported platform.

### CODE REVIEW

**Review by ceolin — APPROVED** [review]

**Comment by ceolin** [comment]
The proposed solution looks good to me, my only concern is if spi drivers start to handle it in their side.

**Comment by JordanYates** [comment]
> The proposed solution looks good to me, my only concern is if spi drivers start to handle it in their side.

As mentioned in point 2, AFAICT any attempt to have SPI drivers handle this would require them doing a lot more work and integrating with several other API's, to handle what is a power domain problem. I don't think it's likely anyone want to pollute their SPI driver with all that.

**Comment by JordanYates** [comment]
> Could you provide an example of this being used in-tree? some board which needs it perhaps?

From the PR description: `The BMI270 sensor on the Nordic Thingy53 is an example of this hardware setup.`

> Also not getting why the follower ones ignore the polarity flag?

Ignoring the polarity is sort of the point. CS by default is active-low, but when the device on the bus literally has no power, the CS pin being high is a bad thing.

> I'm finding it hard to follow why enable-gpios could not just be extended like this:

It probably could, but the solution would basically end up looking like this PR anyway. I added it as a separate property to hopefully make it clearer when the property should be used. Having N enable GPIOs for a power domain doesn't really make sense from the hardware perspective.

**Comment by bjarki-andreasen** [comment]
> > Could you provide an example of this being used in-tree? some board which needs it perhaps?
> 
> From the PR description: `The BMI270 sensor on the Nordic Thingy53 is an example of this hardware setup.`

I meant using follower GPIOs as part of this PR, for example, adding it to the Thingy53.

> > Also not getting why the follower ones ignore the polarity flag?
> 
> Ignoring the polarity is sort of the point. CS by default is active-low, but when the device on the bus literally has no power, the CS pin being high is a bad thing.

My thinking is the SPI peripheral and power domain use the same GPIO, but the configuration/flags are not the same:

```
pd {
        enable-gpios = <gpio0 1 GPIO_ACTIVE_HIGH> /* Switch pin */
                       <gpio0 0 GPIO_ACTIVE_HIGH> /* CS pin */
};

spi {
        power-domain = <&pd>;
        cs-gpios = <gpio0 0 GPIO_ACTIVE_LOW> /* same CS pin */
};
```

And when the power domain is suspended, the SPI must already be suspended, so there is no contention for ownership of the CS GPIO, same goes the other direction, PD is resumed before SPI, and SPI takes ownership over the CS pin when its resumed.

> > I'm finding it hard to follow why enable-gpios could not just be extended like this:
> 
> It probably could, but the solution would basically end up looking like this PR anyway. I added it as a separate property to hopefully make it clearer when the property should be used. Having N enable GPIOs for a power domain doesn't really make sense from the hardware perspective.

I would describe what this PR is doing as doing that in spirit:
```
pd {
        enable-gpios = <gpio0 1 GPIO_ACTIVE_HIGH>; /* Switch pin */
        follower-gpios = <gpio0 0 0>; /* CS pin */
};
```
just with one of the properties ignoring the polarity.

I believe the best option is to have drivers manage their GPIOs, and not share ownership of pins between PD and device drivers. It provides full flexibility to device drivers and its clear for users to understand I think.

**Comment by JordanYates** [comment]
> I believe the best option is to have drivers manage their GPIOs, and not share ownership of pins between PD and device drivers. It provides full flexibility to device drivers and its clear for users to understand I think.

It sounds like you are proposing alternate solution 1 from the PR description. The problem here is completely unrelated to individual device drivers since it is a PCB design problem. Providing a way to solve this issue for all the devices that it could affect would therefore require modifications to every SPI device driver in tree. That sounds unreasonable to me. The problem is restricted to certain power domain setups, which is why I have attached the solution to the power domain.

**Comment by bjarki-andreasen** [comment]
> > I believe the best option is to have drivers manage their GPIOs, and not share ownership of pins between PD and device drivers. It provides full flexibility to device drivers and its clear for users to understand I think.
> 
> It sounds like you are proposing alternate solution 1 from the PR description. The problem here is completely unrelated to individual device drivers since it is a PCB design problem. Providing a way to solve this issue for all the devices that it could affect would therefore require modifications to every SPI device driver in tree. That sounds unreasonable to me. The problem is restricted to certain power domain setups, which is why I have attached the solution to the power domain.

The problem of back powering is quite common in many devices, like interrupts pins from sensors and RTCs, any IO pin really, and the optimal GPIO state after powered off can be quite varied, disconnected? pulled down, set to 1 for I2C bus pins?

See this addition from an input driver:
https://github.com/zephyrproject-rtos/zephyr/blob/302683ef7f21284c3ea90b8a24723ff56d536823/drivers/input/input_gpio_keys.c#L228-L249

I don't believe adding the `follower-gpios` is a flexible enough solution to add compared to the larger problem I believe CS pins hall into. What if we wanted to disconnect the CS pin rather than set it low for example?

If we really want to solve this in a generic way, which I think we do, we need something akin to pinctrl states, to define how each pin should be configured when a device is suspended and/or powered off. It would make sense to have it be part of the power domain infrastructure given the states follow the power domain as this PR proposes, but I think we need to ability to configure the pins behaviors.

**Comment by bjarki-andreasen** [comment]
> > The problem of back powering is quite common in many devices, like interrupts pins from sensors and RTCs, any IO pin really, and the optimal GPIO state after powered off can be quite varied, disconnected? pulled down, set to 1 for I2C bus pins? (and definately varied by the specific board/hw the device is on)
> 
> Correct, and all of these pins you mention (except the I2C bus pins) are specific to the driver, and can (and should!) be set to `GPIO_DISCONNECTED` inside the `PM_DEVICE_ACTION_TURN_OFF` action.

So are CS pins? Why move logic specific to them to the power domain, but not others?

> I don't see the provided snippet as being relevant as it has nothing to do with power domains (suspended vs active is separate from on vs off).

The important part of that snippet is the option: `no-disconnect` added to the devicetree to specify how to configure the pins from the driver. Necessary since GPIO pins vary in features and optimal pin states vary by hardware layout.

> > set to 1 for I2C bus pins
> 
> That is already the default idle state of the bus, nothing needs to be done.
> 
> > What if we wanted to disconnect the CS pin rather than set it low for example?
> 
> That sounds like a standard CS configuration (e.g. `cs-gpios = <&gpio0 21 (GPIO_ACTIVE_LOW | GPIO_OPEN_DRAIN)>;`), in which case the pin will naturally return to 0V when the power domain the pull-up is connected to is turned off.

These are valid configurations of a CS pin:
* The GPIO pin is on the other side of a level shifter which inverts the signal
* The GPIO pin is inverted
* The GPIO pin supports disconnect, but not open-drain
* The GPIO pin supports open-drain, but not disconnect
* The GPIO pin supports open-drain and disconnect, but for this specific hardware, disconnect is more efficient than open-drain.

Simply setting it low is not nearly enough to cover ll of these cases.

> > If we really want to solve this in a generic way, which I think we do, we need something akin to pinctrl states, to define how each pin should be configured when a device is suspended and/or powered off.
> 
> I don't see that solution as being possible. If dev A requests pin X to be high when its off, and dev B requests pin X to be low when off, and both are off...

Do you have an example where this is the case? devices don't usually share the same GPIO, and if they do, it requires quite a bit of synchronizing between the two drivers.

The reason PD is a good place to put alternate configurations is that when devices are off, they don't control the GPIOs, so the PD can take ownership and configure them.

**Comment by kartben** [comment]
What's the status of this PR? It seems to be missing one approval, but I don't see any request for changes either?

**Review by bjarki-andreasen — CHANGES_REQUESTED** [review]
I don't think this is the correct approach, see previous comments

**Comment by JordanYates** [comment]
> So are CS pins? Why move logic specific to them to the power domain, but not others?

Because CS pins are generally not touched by device drivers, they are used behind the scenes by the SPI API.

> These are valid configurations of a CS pin:
    1. The GPIO pin is inverted
    2. The GPIO pin supports disconnect, but not open-source
    3. The GPIO pin supports open-source, but not disconnect
    4. The GPIO pin supports open-source and disconnect, but for this specific hardware, disconnect is more efficient than open-source with value 0.

1. The GPIO pin is already at 0V when idle, no need to use this feature
2-4. This feature does not attempt to configure anything as open-source or disconnect, not sure how it is relevant.

> Simply setting it low is not nearly enough to cover ll of these cases.

On the contrary, setting it low is the sole use-case this feature is attempting to address. "A GPIO that normally idles as a positive voltage, which should be driven to 0V when the power domain is powered down".

> I don't think this is the correct approach, see previous comments

I'm not entirely clear on what you think the correct approach is.
Is it moving the logic somewhere else (SPI drivers, individual device drivers)?
Is it moving the pin specification into the `enable-gpios` property?
It it that this feature should be able to configure a GPIO into any arbitrary state when the power domain goes down?

There seems to be a lot of points about GPIO's being configured as `disconnect` or `open-source`, but I don't see the relevance to the problem that is trying to be solved.

**Comment by bjarki-andreasen** [comment]
> 1 The GPIO pin is already at 0V when idle, no need to use this feature
> 2-4 This feature does not attempt to configure anything as open-source or disconnect,
> not sure how it is relevant.

I get that, I'm saying it has to to be generally useful. CS pins, like any other pin, can be configured in any of these configurations validly, given some intermediary circuit between the host CS pin and the device CS pin, like a level shifter, inverter, isolator etc. A CS pin will not always be connected directly from host to device.

Expanding on that, when it comes to back powering or leakage of any kind, different GPIO configurations which result in the "same" output, can result in different current draw depending on the host, so configuring a pin as an input, disconnected if that is supported, or setting it as an output low, has to be configurable.

This PR does not address most of these real problems, only the case where setting it output low happens to be the correct configuration when power domain is off.

Why not expand this PR to support all these cases, including cases like configuring a potential IRQ GPIO and whatever other GPIOs are connected to the device to acheive the lowest possible leakage when power domain is off?

So the problem with this PR, summarized, is that it only handles one subset of the larger issue of "back powering" / leakage, which I believe can and should be solved as a whole, which would cover the configuration this PR supports

**Comment by JordanYates** [comment]
> Why not expand this PR to support all these cases, including cases like configuring a potential IRQ GPIO and whatever other GPIOs are connected to the device to acheive the lowest possible leakage when power domain is off?

Because all of those GPIOs are unique (in that not every device has IRQ1, IRQ2, etc, etc) to the instantiation of the device driver, and the logic for configuring those GPIOs is best left in the driver.

The CS pin is different in that every (more or less) SPI device has one, and yet no drivers (more or less) touch the CS pin at any point in the driver code. I agree that the logic *could* be handled in individual device drivers. I disagree that it is a good idea, since you are inserting workarounds for hardware design errors (no external pull-up on the CS line to the voltage rail) in >100 in-tree drivers, rather than at the source of the problem (the power domain).

To cover the hardware situation where the CS control has an inline inverter (for some reason) AND the other side doesn't use pull-ups properly, I could add a way to make the raw follower go physical high instead when the power domain goes down. Would that be sufficient?

**Comment by bjarki-andreasen** [comment]
@JordanYates Taking a step back from the power domain, I found this function `spi_context_cs_control()` which directly controls the CS line, it is used quite a lot across SPI drivers. Are you certain that: 

> The CS pin is different in that every (more or less) SPI device has one, and yet no drivers (more or less) touch the CS pin at any point in the driver code

is accurate? Is it really that much of an ask to add an additional call to the common spi context like `spi_context_cs_control_raw` and calling it from the pm action callback on off in the drivers?

**Comment by JordanYates** [comment]
> @JordanYates Taking a step back from the power domain, I found this function `spi_context_cs_control()` which directly controls the CS line, it is used quite a lot across SPI drivers. Are you certain that:

It is used in quite a lot (all?) *SPI* interface drivers, and 0 SPI device drivers. SPI interface drivers are not the right place to put this logic for the reasons in the original PR description.

**Comment by bjarki-andreasen** [comment]
> > @JordanYates Taking a step back from the power domain, I found this function `spi_context_cs_control()` which directly controls the CS line, it is used quite a lot across SPI drivers. Are you certain that:
> 
> It is used in quite a lot (all?) SPI interface drivers, and 0 SPI device drivers. SPI interface drivers are not the right place to put this logic for the reasons in the original PR description.

Right, my mistake :)
